geospaas_processing/copiers.py

```diff
@@ -0,0 +1,72 @@
+import os
+import shutil
+import logging
+
+import django
+
+from geospaas.catalog.models import Dataset
+from geospaas.catalog.managers import LOCAL_FILE_SERVICE
+
+os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'geospaas_processing.settings')
+django.setup()
+
+LOGGER = logging.getLogger(__name__)
+
+
+class Copier():
+    """Copier for datasets"""
+
+    def __init__(self, type_in_flag_file, destination_path, flag_file_request=False,
+                 link_request=False, **criteria):
+        self.type_in_flag_file = type_in_flag_file
+        self.destination_path = destination_path
+        self.flag_file_request = flag_file_request
+        self.link_request = link_request
+        self.datasets = Dataset.objects.filter(**criteria)
+
+    @staticmethod
+    def write_flag_file(type_in_flag_file, source_path, dataset, sopposed_full_address):
+        """
+        writes the flag file in the case of request for it. ".flag" is the extension for flag file.
+        """
+        with open(sopposed_full_address + ".flag", "w") as flag_file:
+            flag_file.write(f"type: {type_in_flag_file}" + os.linesep)
+            for urlname in dataset.dataseturi_set.exclude(service=LOCAL_FILE_SERVICE):
+                flag_file.write(f"url: {urlname.uri}" + os.linesep)
+
+    def file_or_symlink_copy(self, source_paths, dataset):
+        """copy the file or a symlink of the file of dataset based on its stored local address in
+        the database."""
+        for source_path in source_paths:
+            if os.path.isfile(source_path.uri):
+                sopposed_full_address = os.path.join(self.destination_path.rstrip(os.path.sep),
```

Comment:
```suggestion
                destination_filename = os.path.join(self.destination_path,
```
1. please rename `sopposed_full_address` everywhere
2. `os.path.join('a', 'b')` and `os.path.join('a/', 'b')` produce the same result. That what `join` is for.

Reply:
Please have a look:
![image](https://user-images.githubusercontent.com/62345618/100077252-6f012f80-2e42-11eb-921b-eb59739af92c.png)

The third and the fourth case are the cases that I am worried about! It is better to have these rstrip and also lstrip in the code in order not to deal with this simple issue later on!

Reply:
The name has been changed. Done in the next commit.

Reply:
I agree with Anton on the second point.

`a//b` is equivalent to `a/b`.
`/b` is an absolute path, which should only be used as first argument to `os.path.join()`. This situation will never happen because you use `os.path.basename()` on the second argument.

Reply:
Done in the next commit.